src/utils/wallets/wallets.utils.ts

```diff
@@ -185,77 +176,147 @@ async function generateWalletJWKFromShares(
+  shareHash: string;
+  sharePublicKey: string;
 }
 
-// Data (localStorage):
+/**
+ * See:
+ * - https://github.com/framp/zero-knowledge-node/blob/master/keypair-auth/crypto-utils.js
+ * - https://www.youtube.com/watch?v=cMoD0wIxIpQ
+ */
+async function generateShareHashAndPublicKey(
+  share: string
+): Promise<GenerateShareHashAndPublicKeyReturn> {
+  log(LOG_GROUP.WALLET_GENERATION, "generateShareHashAndPublicKey()");
+
+  const sharePrng = random.createInstance();
+
+  sharePrng.seedFileSync = (needed) => {
+    let r = "",
+      i = 0,
+      j = 0;
+
+    while (i++ < needed) {
+      r += share[j++];
+      if (j === share.length) j = 0;
+    }
 
-// Device Nonce:
+    return r;
+  };
 
-const DEVICE_NONCE_KEY = "DEVICE_NONCE";
+  return new Promise((resolve, reject) => {
+    pki.rsa.generateKeyPair(
+      {
+        bits: 4096,
+        prng: sharePrng
+      },
+      async (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          const shareHash = await generateShareHash(share);
+          const publicKey = result.publicKey;
+          const publicKeyPEM = pki.publicKeyToPem(publicKey);
+          const sharePublicKey = pemToBase64(publicKeyPEM);
+
+          resolve({
+            shareHash,
+            sharePublicKey
+          });
+        }
+      }
+    );
+  });
+}
 
-export type DeviceNonce =
-  `${number}-${number}-${number}T${number}:${number}:${number}.${number}Z-${string}`;
+export interface GenerateShareHashAndPrivateKeyReturn {
+  shareHash: string;
+  sharePrivateKeyJWK: JWKInterface;
+}
 
-function loadDeviceNonce(): DeviceNonce | null {
-  let deviceNonce = localStorage.getItem(DEVICE_NONCE_KEY) || null;
+/**
+ * See:
+ * - https://github.com/framp/zero-knowledge-node/blob/master/keypair-auth/crypto-utils.js
+ * - https://www.youtube.com/watch?v=cMoD0wIxIpQ
+ */
+async function generateShareHashAndPrivateKey(
+  share: string
+): Promise<GenerateShareHashAndPrivateKeyReturn> {
+  log(LOG_GROUP.WALLET_GENERATION, "generateShareHashAndPrivateKey()");
+
+  const sharePrng = random.createInstance();
+
+  sharePrng.seedFileSync = (needed) => {
+    let r = "",
+      i = 0,
+      j = 0;
+
+    while (i++ < needed) {
+      r += share[j++];
+      if (j === share.length) j = 0;
+    }
 
-  if (
-    deviceNonce === null ||
-    /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)\-[\w_-]{21}/.test(
-      deviceNonce
-    )
-  )
-    return deviceNonce as DeviceNonce;
+    return r;
+  };
 
-  if (process.env.NODE_ENV === "development") {
-    throw new Error(INVALID_DEVICE_NONCE_ERR_MSG);
-  } else {
-    console.warn(INVALID_DEVICE_NONCE_ERR_MSG);
-  }
+  return new Promise((resolve, reject) => {
+    pki.rsa.generateKeyPair(
+      {
+        bits: 4096,
+        prng: sharePrng
+      },
+      async (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          const shareHash = await generateShareHash(share);
```

Comment:
@7i7o I'll probably open a GitHub issue on `node-forge` to see if I get some help on this while I work on other things. The hash-based signatures are working anyway, so that should be enough for now.